### DISCUSSION TITLE
How to properly update links in collection

### DISCUSSION THREAD

**Discussion by intik** [isar/isar#71] [Help]
**Steps to reproduce**
I have two simple collections, like in the docs: Student, Teachers 
Student have field teachers of links to teachers.
I want to have simple form where i can edit Student and his teachers at one.
How to properly save changed teachers (new ones and some edited) along with student edit?

**Simplified code**
```
@Collection class Student { 
 String name;
 @IsarLinks teachers = IsarLinks<Teacher>();
}

@Collection class Teacher {
 String name;
}

final student = await isar.students.get(1);
await student.teachers.load();

student.name = 'Luke';
student.teachers.elementAt(1).name = 'Joe';
student.teachers.add(Teacher()..name='April');

isar.writeTxn((isar) async {
 await isar.students.put(student);
 await student.teachers.saveChanges();
 // How to save changed student, added teacher and changed teacher?
};

```
**Issue**
I expected that the solution above would work, but no somehow existing teachers always dissapears from the student. They remain in the db (I can query for them explicitly), but newly loaded (refreshed) student only contains the newly added teacher.

I've tried different combinations of what to when save (clearing teachers from student and adding again and more). Closest was when I manually copied every teacher to a new object, clear student teachers and and new copies of teachers, which is not optimal 😀 I see in code, that `saveChanges()` probably only applies to newly added/removed items in list, I would be ok that existing teachers wasn't updated, but why are the removed from studen teachers list?

Am I missing something?

**Version**
 - Platform: Windows
 - Flutter version: 2.0.1
 - Isar version: 0.4.0
 - Isar Generator version: 0.4.0

**Comment by simc** [comment]
@intik The `.saveChanges()` method only saves changes you made to the link (adding or removing linked objects). The following code should work:

```dart
student.name = 'Luke';
student.teachers.elementAt(1).name = 'Joe';
student.teachers.add(Teacher()..name='April');

isar.writeTxn((isar) async {
 await isar.students.put(student);
 await student.teachers.saveChanges();
 await isar.teachers.put(student.teachers.elementAt(1));
};
```

**Reply by richard457** [reply]
@leisim  .saveChanges(); is no longer available is this same as save()?

**Reply by richard457** [reply]
```
 if (item != null) {
      order.orderItems.add(item);
      return isar.writeTxn((isar) async {
        return order.orderItems.save();
      });
    }
```
I am using the above snippet, where I am providing the existing item in assumption that it will update but it is not working.

**Reply by simc** [reply]
@richard457 and the object is already stored in isar?

**Reply by richard457** [reply]
Yes, but could not be updated. @leisim for the adding object it work fine. @leisim

**Comment by ch-muhammad-adil** [comment]
Hello @leisim , In my case I am linking data of same type. 

I have following setup

```
Category {
   @ignore
   List<Category> childrenCategories;
   final childrenLinks= IsarLinks<Category>();
}
```

and this is not working 

  ```
await _isar!.writeTxn(() async {
      await _isar!.categorys.putAll(categories);
      categories.forEach((element) {
        element.childrenLink.save();
      });
    });
```

I am using .save() method and .saveChanges() is not available. 
Everything else is working but the links data is empty. Any comment?

**Comment by ch-muhammad-adil** [comment]
The above case wasn't working for me, after some R&D below case has worked. I am first saving main list of categories then I loop on categories list and save parents/children IsarLinks in another transaction(see above code👆). 

WORKING 👍
```
    await _isar!.writeTxn(() async {
      await _isar!.categorys.putAll(categories);
    });
    await _isar!.writeTxn(() async {
      categories.forEach((element) async {
        await element.childrenLink.save();
        await element.parentsLink.save();
      });
    });
```

SEEMS FUNNY::: I tried to save IsarLinks with below code(inside same transaction), but it has caused crash every time I run application. I tested this case for like 5 times just to confirm. With above code it works good.  I think so logic is first save the main objects and then create IsarLinks between those objects. 

 ```
    await _isar!.writeTxn(() async {
      await _isar!.categorys.putAll(categories);
      categories.forEach((element) async {
        await element.childrenLink.save();
        await element.parentsLink.save();
      });

    });
```
